Improve support for Tuya smoke sensor `_TZE284_rccxox8p`

sarxos opened the pull request:
This commit extends existing support for Tuya TS0601 _TZE284_rccxox8p by covering:

* Smoke detection sensor
* Smoke concentration sensor
* Mute configuration
* Battery level diagnostic
* Sensor failure diagnostic
* Self-test diagnostic

Previously, the only sensor available was:

* Smoke detection sensor

## Proposed change

Before:

![image](https://github.com/user-attachments/assets/21bc3d07-99c7-4425-aa8e-1e3080c094df)

After:

![image](https://github.com/user-attachments/assets/b752bd4a-d15e-42a2-b81d-697821d7d9e9)

## Additional information

Device looks like this:

![image](https://github.com/user-attachments/assets/402daa5d-6ab9-4c02-914d-0756cc3f4a23)

Information about Tuya DPs exposed by this device has been found in https://github.com/Koenkk/zigbee-herdsman-converters/blob/a19bc8c4652496998f32ac777963379f9c7c24cf/src/devices/tuya.ts#L6553

I own the device. Have tested it with my Nabu Casea Home Assistant Yellow with ZHA, core 2025.4.3, by prototyping custom quirk first and then changing the original quirk file directly in the homeassistant container in:

```
/usr/local/lib/python3.13/site-packages/zhaquirks/tuya/tuya_smoke.py
```

Tested functionalities:

* Smoke detection - pass
* Mute swicth when alarm is triggered - pass
* Self-test button - pass

Not tested:

* Sensor failure - obviously, would like to avoid having my device damaged :D
* Battery level - didn't find any older batteries at home, have only 100% new ones, so the battery was always 100% when playing with the device

Here are the values when smoke is detected:

![image](https://github.com/user-attachments/assets/314cb40e-fa9d-4f54-afa5-5ab155ff8273)

Mute works both ways. One can change it from configuration panel or directly by pressing the button on the device.

![image](https://github.com/user-attachments/assets/37c8c5e3-aa7c-4c7b-b529-663f3a181d0c)

It may be worth mentioning that self-test won't change the smoke detection sensor but, instead, a different sensor designed just for test purposes. Integrators in HA should keep this in mind. I was a bit surprised but it is what it is.

Self-test sensor:

![image](https://github.com/user-attachments/assets/dfa4f6f1-4b93-4b25-a77b-c657b5a99129)

Smoke detection event in history:

![image](https://github.com/user-attachments/assets/5ae358f5-0427-4d42-ba65-fce3b9c90b81)

## Checklist

- [x] The changes are tested and work correctly
- [ ] `pre-commit` checks pass / the code has been formatted using Black
- [ ] Tests have been added to verify that the new code works



Closes #4687
Closes #4546
Related: #4212

sarxos committed:
Add full support for Tuya TS0601 _TZE284_rccxox8p

This commits extends existing support for Tuya TS0601 _TZE284_rccxox8p
by adding:

* Smoke concentration sensor
* Mute configuration
* Battery level diagnostic
* Sensor failure diagnostic
* Self-test diagnostic

ashmckenzie commented:
Amazing work @sarxos !!! 

I've tested this locally in my setup (HA OS) and works perfectly :+1: 

I initially tried to copy `zhaquirks/tuya/tuya_smoke.py` into my custom ZHA quirks folder, but that didn't work. Do you happen to have a copy of the custom quirk you used?

Regardless, editing `/usr/local/lib/python3.13/site-packages/zhaquirks/tuya/tuya_smoke.py` directly and restarting HA worked!

ashmckenzie commented:
Just a quick one to note that after a few days the battery levels are no longer being reported? Do you see the same behaviour @sarxos ?

TheJulianJES commented on zhaquirks/tuya/tuya_smoke.py:
```diff
@@ -63,6 +66,9 @@ def _update_attribute(self, attrid, value):
             )
 
 
+CONCENTRATION_PARTS_PER_MILLION: Final = "ppm"
```
We should be able to remove this and import it from zigpy:
```python
from zigpy.quirks.v2.homeassistant import CONCENTRATION_PARTS_PER_MILLION
```

TheJulianJES commented on zhaquirks/tuya/tuya_smoke.py:
```diff
@@ -76,16 +82,59 @@ def _update_attribute(self, attrid, value):
+        dp_id=2,
+        type=t.int16s,
+        attribute_name="smoke_concentration",
+        translation_key="smoke_concentration",
+        fallback_name="Smoke concentration",
```
Minor thing, but just to be consistent with other v2 quirks, can we move `translation_key` and `fallback_name` to the bottom in all v2 quirk methods, so below `unit`.

TheJulianJES commented on zhaquirks/tuya/tuya_smoke.py:
```diff
@@ -76,16 +82,59 @@ def _update_attribute(self, attrid, value):
+        fallback_name="Smoke concentration",
+        device_class=SensorDeviceClass.PM10,
+        state_class=SensorStateClass.MEASUREMENT,
+        converter=lambda value: value / 10,
```
Also minor: we should be able to use `divisor=10` here instead.

TheJulianJES commented on zhaquirks/tuya/tuya_smoke.py:
```diff
@@ -76,16 +82,59 @@ def _update_attribute(self, attrid, value):
+    .applies_to("_TZE200_m9skfctm", "TS0601")
+    .applies_to("_TZE200_rccxox8p", "TS0601")
+    .tuya_smoke(dp_id=1)
+    .tuya_sensor(
+        dp_id=2,
+        type=t.int16s,
+        attribute_name="smoke_concentration",
+        translation_key="smoke_concentration",
+        fallback_name="Smoke concentration",
```
Depending on how ppm is reported for this device, we could even add `suggested_display_precision=0` to remove the decimal numbers at the end.

We might need to bump the zigpy version if tests fail, but I'll do that then.

TheJulianJES replied:
See the comments from the review bot below. It made this possible in the first place but it depends on what the device sends if we actually want this.

TheJulianJES reviewed (commented):
Otherwise, I think this looks good. Thanks!
Nice effort for the PR description 😄

ashmckenzie commented:
@sarxos I'd really like to see this merged and am happy to make the suggested changes if you're at capacity?

zigpy-review-bot committed:
Merge remote-tracking branch 'origin/dev' into TS0601-TZE284-rccxox8p-full

zigpy-review-bot committed:
Forward `suggested_display_precision` through `tuya_sensor`

zigpy-review-bot committed:
Address review feedback on `_TZE284_rccxox8p` smoke quirk

- Import `CONCENTRATION_PARTS_PER_MILLION` from zigpy instead of defining it locally
- Use `divisor=10` instead of a converter lambda for smoke concentration
- Drop `device_class=SensorDeviceClass.PM10` (HA requires µg/m³ for PM10, not ppm)
- Add `suggested_display_precision=0` for the ppm sensor
- Move `translation_key`/`fallback_name` to the end of each v2 entity call

zigpy-review-bot committed:
Drop leading underscore from self_test attribute name

zigpy-review-bot committed:
Use `fault_alarm` naming for device fault sensor

Matches the existing ZHA `strings.json` `fault_alarm` translation and
sibling Tuya quirks (tuya_gas, tuya_thermostat), so the entity name
translates upstream and `translation_key`/`fallback_name` stay consistent.
